clippy_lints/src/non_send_field_in_send_ty.rs

```diff
@@ -0,0 +1,129 @@
+    ///
+    /// #[derive(Debug, Clone)]
+    /// pub struct ArcGuard<RC, T> {
+    ///     inner: T,
+    ///     // Possibly drops `Arc<RC>` (and in turn `RC`) on a wrong thread
+    ///     head: Arc<RC>
+    /// }
+    /// ```
+    pub NON_SEND_FIELD_IN_SEND_TY,
+    nursery,
+    "there is field that does not implement `Send` in a `Send` struct"
+}
+
+declare_lint_pass!(NonSendFieldInSendTy => [NON_SEND_FIELD_IN_SEND_TY]);
+
+impl<'tcx> LateLintPass<'tcx> for NonSendFieldInSendTy {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'_>) {
+        // Checks if we are in `Send` impl item.
+        // We start from `Send` impl instead of `check_field_def()` because
+        // single `AdtDef` may have multiple `Send` impls due to generic
+        // parameters, and the lint is much easier to implement in this way.
+        if_chain! {
+            if let Some(send_trait) = cx.tcx.get_diagnostic_item(sym::send_trait);
+            if let ItemKind::Impl(hir_impl) = &item.kind;
+            if let Some(trait_ref) = &hir_impl.of_trait;
+            if let Some(trait_id) = trait_ref.trait_def_id();
+            if send_trait == trait_id;
+            if let ImplPolarity::Positive = hir_impl.polarity;
+            if let Some(ty_trait_ref) = cx.tcx.impl_trait_ref(item.def_id);
+            if let self_ty = ty_trait_ref.self_ty();
+            if let ty::Adt(adt_def, impl_trait_substs) = self_ty.kind();
+            then {
+                for variant in &adt_def.variants {
+                    for field in &variant.fields {
+                        let field_ty = field.ty(cx.tcx, impl_trait_substs);
+
+                        if raw_pointer_in_ty_def(cx, field_ty)
+                            || implements_trait(cx, field_ty, send_trait, &[])
+                            || is_copy(cx, field_ty)
+                        {
+                            continue;
+                        }
+
+                        if let Some(field_hir_id) = field
+                            .did
+                            .as_local()
+                            .map(|local_def_id| cx.tcx.hir().local_def_id_to_hir_id(local_def_id))
+                        {
+                            if let Some(field_span) = cx.tcx.hir().span_if_local(field.did) {
+                                span_lint_hir_and_then(
+                                    cx,
+                                    NON_SEND_FIELD_IN_SEND_TY,
+                                    field_hir_id,
+                                    field_span,
+                                    "non-`Send` field found in a `Send` struct",
+                                    |diag| {
+                                        diag.span_note(
+                                            item.span,
+                                            &format!(
+                                                "type `{}` doesn't implement `Send` when `{}` is `Send`",
```

Comment:
The "when" statement seems a bit weird, as the type restriction is not conditional, from my understanding.

I would structure the error message to first show the item span in the `span_lint` with the error message and then add additional information. Making it look like this is more direct and the user knows where this originated from:

```
error: this implementation is unsound, as some fields in `MvccRwLock<T>` are `!Send`
  --> $DIR/non_send_field_in_send_ty.rs:23:1
   |
LL | unsafe impl<T> Send for MvccRwLock<T> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the field type of `lock` doesn't implement `Send`
  --> $DIR/non_send_field_in_send_ty.rs:20:5
   |
LL |     lock: Mutex<Box<T>>,
   |     ^^^^^^^^^^^^^^^^^^^
   |
   = help: use a type that implements `Send`
   = help: or consider adding a `T: Send` bound
```
I've also changed the lint messages in the example to be a bit more direct. The second help text should only be added when the type actually uses a generic. You can use `clippy_utils::is_lint_allowed(cx, NON_SEND_FIELD_IN_SEND_TY, field_hir_id)` to check for `#[allow]` attributes on the field itself.

Reply:
> The "when" statement seems a bit weird, as the type restriction is not conditional, from my understanding.

Trait bound on type parameters may make `Send` impl conditional:
```rust
struct MyStruct {
    field: T,
}

unsafe impl<T: MyTrait> Send for MyStruct<T> {}
```

What I wanted to express was that `T: MyTrait` precondition (= "when `MyStruct<T>` is `Send`") does not guarantee that `T: Send`.

With that being clarified, I really like your suggestion of the lint message. I think specifying the unsafe impl first and saying "this implementation is unsound" provides a clear context for the users.